doc/developer/design/20231113_optimizer_notice_catalog.md

```diff
@@ -0,0 +1,555 @@
+   [`mz_optimizer_notices`](#mz_optimizer_notices).
+4. The proposed [SQL syntax extensions](#sql-syntax-extensions).
+
+We will need to cover at least (1) before we can enable the feature for internal
+use.
+
+### RBAC handling
+
+Currently, `Builtin` objects are created with a `sensitivity` field that
+is assigned a `DataSensitivity` value.
+
+```rust
+/// The extent to which data in a builtin object
+/// should be considered "sensitive" and therefore
+/// access to it restricted.
+#[derive(Clone, Debug, Hash, Serialize)]
+pub enum DataSensitivity {
+    /// Any user may query the object.
+    Public,
+    /// Superusers or Materialize staff may query the object.
+    SuperuserAndSupport,
+    /// Only superusers may query the object.
+    Superuser,
+}
+```
+
+Some of the builtin objects proposed in the [Catalog Schema](#catalog-schema)
+(such as `mz_all_optimizer_notices`) are only meant to store raw data and
+support the construction of user-facing builtin views. Those can be all defined
+with sensitivity `DataSensitivity::Superuser`.
+
+For the MVP, we decided to go with a combination of a `Superuser` and
+`SuperuserAndSupport` policies where:
+
+- `mz_all_optimizer_notices` uses `Superuser`.
+- `mz_all_optimizer_notices_redacted` uses `Public`. The view is redacted by
+  re-binding `~` to `~_redacted`.
+
+This means that.
+
+- ☑ All notices will be only to superusers.
+- ☑ All notices will be only to Materialize support engineers.
+- ☑ The design is SOC2 compliant.
+- ☒ Regular users and will see redacted hints by default and might require
+  assistance from a superuser if there are ambiguities.
+
+Discussed Alternatives can be found in [the alternatives
+section](#rbac-handling-alternatives).
+
+<!--
+### Notes:
+
+- `Catalog::open` returns a `builtin_table_updates` vector.
+- The vector is then moved to the `Coordinator::bootstrap` call.
+- The `bootstrap` method than calls `Coordinator::send_builtin_table_updates_blocking`.
+  - Luckily for us this happens after the `for entry in &entries` loop.
+  - Call sites:
+    - `Coordinator::bootstrap`.
+    - `Coordinator::catalog_transact_with`.
+    - `Coordinator::add_active_subscribe`.
+    - `Coordinator::remove_active_subscribe`.
+- The code in `builtin_table_updates.rs` handles update generation.
+  - For example see `CatalogState::pack_materialized_view_update`.
+    - Called by `CatalogState::pack_item_update`.
+      - Called in `Catalog::transact_inner`.
+      - Called in `Catalog::transact_alter_set_cluster`.
+      - Called in `Catalog::open`.
+- Should the design be i18n-ready?
+  - We cannot do that at the moment because we need to write fully-resolved notices to the tables.
+- When do we call `OptimizerNotice::to_string()`?
+  - When writing in the `mz_internal` table?
+    - We are kind of forced to do it like that at the moment.
+  - When reading from the `mz_internal` table?
+    - This can be quite tricky to achieve.
+-->
+
+## Alternatives
```

Comment:
What about durably storing the notices, or some portion of the notice, in the Stash, instead of a builtin table/source? This is what we do for object comments, and it might alleviate some of the problems around retractions?

Reply:
> What about durably storing the notices, or some portion of the notice, in the Stash, instead of a builtin table/source? 

I think the main reason not to do that is that that optimizer notices are a function of the catalog state, more specifically a lot of them would be a function of properties of the plans for materialized views and indexes created during bootstrap.

This means that when we restart / upgrade materialize the plans might change. We would need to somehow discard all optimizer notices created before the restart and create new set of notice entries for the builtin source. Given these semantics, I thought that maintaining the optimizer notices in a builtin table is sufficient for now.